diff -u a/btorexp.c b/btorexp.c
--- a/btorexp.c	2011-03-04 19:49:37.000000000 +0100
+++ b/btorexp.c	2011-03-24 14:11:59.454945556 +0100
@@ -27,7 +27,7 @@
 #include "btorconfig.h"
 #include "btorexit.h"
 #include "btorrewrite.h"
-#include "../picosat/picosat.h"
+#include <picosat.h>
 
 #include <stdlib.h>
 #include <stdio.h>
diff -u a/btorsat.c b/btorsat.c
--- a/btorsat.c	2010-06-14 18:38:20.000000000 +0200
+++ b/btorsat.c	2011-03-24 14:12:33.307862210 +0100
@@ -17,7 +17,7 @@
  *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#include "../picosat/picosat.h"
+#include <picosat.h>
 
 #ifdef BTOR_USE_PICOPREP
 #include "../picoprep/picoprep.h"
