Index: scripts/Makefile.common
===================================================================
--- scripts/Makefile.common	(revision 1658)
+++ scripts/Makefile.common	(working copy)
@@ -16,7 +16,7 @@
 endif
 
 # Optimization
-OPTIMIZE      = -O3 -DNDEBUG -fomit-frame-pointer $(OPTIMIZE_MARCH_NATIVE)
+OPTIMIZE      = -fPIC -O3 -DNDEBUG -fomit-frame-pointer $(OPTIMIZE_MARCH_NATIVE)
 #OPTIMIZE      =  -O3 -g            # Debug
 
 # To enable 128-bit indices in the array propagators. Put the INDICES_128BITS
Index: src/sat/mtl/template.mk
===================================================================
--- src/sat/mtl/template.mk	(revision 1658)
+++ src/sat/mtl/template.mk	(working copy)
@@ -24,7 +24,7 @@
 
 COPTIMIZE ?= -O3
 
-CFLAGS    += -I$(MROOT) -D __STDC_LIMIT_MACROS -D __STDC_FORMAT_MACROS
+CFLAGS    += -fPIC -I$(MROOT) -D __STDC_LIMIT_MACROS -D __STDC_FORMAT_MACROS
 LFLAGS    += -lz
 
 .PHONY : s p d r rs clean 
