From 8102c05204e5e5aba9094d7a51363b68131f4fb3 Mon Sep 17 00:00:00 2001
From: Finn Haedicke <finn@informatik.uni-bremen.de>
Date: Thu, 17 Jun 2010 14:48:47 +0200
Subject: [PATCH 04/14] added ifdefs to handle includes of from C++

---
 boolector.h |    7 +++++++
 btordefs.h  |    8 ++++++++
 2 files changed, 15 insertions(+), 0 deletions(-)

diff --git a/boolector.h b/boolector.h
index 953dd69..e80b5a2 100644
--- a/boolector.h
+++ b/boolector.h
@@ -25,6 +25,9 @@
 /*
 Boolector is copyrighted 2007 - 2009 by Robert Brummayer, Armin
 */
+#ifdef __cplusplus
+extern "C" {
+#endif 
 
 #include "btordefs.h"
 #include <stdio.h>
@@ -1028,4 +1031,8 @@ void boolector_free_bv_assignment (Btor * btor, char * assignment);
  **/
 void boolector_abort_function( void (*fun)(int) );
 
+#ifdef __cplusplus
+} // extern C
+#endif 
+
 #endif
diff --git a/btordefs.h b/btordefs.h
index f02a0b9..19635e7 100644
--- a/btordefs.h
+++ b/btordefs.h
@@ -1,7 +1,15 @@
 #ifndef BTORDEFS_H_INCLUDED
 #define BTORDEFS_H_INCLUDED
 
+#ifdef __cplusplus
+extern "C" {
+#endif 
+
 typedef struct BtorExp BtorExp;
 typedef struct Btor Btor;
 
+#ifdef __cplusplus
+} // extern "C"
+#endif 
+
 #endif
-- 
1.7.3.4

