--- src/btorsat.h	2016-12-17 03:47:54.000000000 +0100
+++ src/btorsat.patch.h	2017-01-20 22:26:09.199798247 +0100
@@ -83,7 +83,7 @@
 };
 
 #if defined(BTOR_USE_LINGELING)
-#include "lglib.h"
+#include <lglib.h>
 typedef struct BtorLGL BtorLGL;
 
 struct BtorLGL
--- src/btorsat.c	2016-12-17 03:47:54.000000000 +0100
+++ src/btorsat.patch.c	2017-01-20 22:25:53.280079507 +0100
@@ -17,7 +17,7 @@
 #include "picosat.h"
 #endif
 #ifdef BTOR_USE_LINGELING
-#include "lglib.h"
+#include <lglib.h>
 #endif
 #ifdef BTOR_USE_MINISAT
 #include "btorminisat.h"
